src/protocol/sender/mod.rs

```diff
@@ -0,0 +1,499 @@
+        }
+
+        //   5) If the sequence number of the current packet is 16n, where n is an
+        //      integer, go to 2).
+        if let Some(p) = self.pop_transmit_buffer_16n() {
+            //      NOTE: to get the closest timing, we ignore congestion control
+            //      and send the 16th packet immediately, instead of proceeding to step 2
+            self.send_data(p);
+        }
+
+        //   6) Wait (SND - t) time, where SND is the inter-packet interval
+        //      updated by congestion control and t is the total time used by step
+        //      1 to step 5. Go to 1).
+        self.step = Step6;
+        let period = self.congestion_control.send_interval();
+        self.snd_timer.set_period(period);
+        WaitUntil(self.snd_timer.next_instant())
+    }
+
+    fn handle_data_packet(&mut self, _packet: DataPacket, _now: Instant) -> SenderResult {
+        Ok(())
+    }
+
+    fn handle_control_packet(&mut self, packet: ControlPacket, now: Instant) -> SenderResult {
+        match packet.control_type {
+            ControlTypes::Ack {
+                ack_seq_num,
+                ack_number,
+                rtt,
+                rtt_variance,
+                packet_recv_rate,
+                est_link_cap,
+                ..
+            } => self.handle_ack_packet(
+                now,
+                ack_seq_num,
+                ack_number,
+                rtt,
+                rtt_variance,
+                packet_recv_rate,
+                est_link_cap,
+            ),
+            ControlTypes::Ack2(_) => {
+                warn!("Sender received ACK2, unusual");
+                Ok(())
+            }
+            ControlTypes::DropRequest { .. } => unimplemented!(),
+            ControlTypes::Handshake(shake) => self.handle_handshake_packet(shake, now),
+            // TODO: reset EXP-ish
+
+            // TODO: case UMSG_CGWARNING: // 100 - Delay Warning
+            //            // One way packet delay is increasing, so decrease the sending rate
+            //            ControlTypes::DelayWarning?
+
+            // TODO: case UMSG_LOSSREPORT: // 011 - Loss Report is this Nak?
+            // TODO: case UMSG_DROPREQ: // 111 - Msg drop request
+            // TODO: case UMSG_PEERERROR: // 1000 - An error has happened to the peer side
+            // TODO: case UMSG_EXT: // 0x7FFF - reserved and user defined messages
+            ControlTypes::Nak(nack) => self.handle_nack_packet(nack),
+            ControlTypes::Shutdown => self.handle_shutdown_packet(),
+            ControlTypes::Srt(srt_packet) => self.handle_srt_control_packet(srt_packet),
+            // The only purpose of keep-alive packet is to tell that the peer is still alive
+            // nothing needs to be done.
+            // TODO: is this actually true? check reference implementation
+            ControlTypes::KeepAlive => Ok(()),
+        }
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    fn handle_ack_packet(
```

Comment:
This is a bit gross. Could you break out the Ack struct enum variant into it's own struct, maybe `AckControlInfo`?

Reply:
I completely agree, yet this is something I am confident I will do to when I lay into the packet module a bit more. I think there are other improvements to be made here. Essentially I'm saying yes, and I will do this later. It will be ugly for now until it gets better, but rest assured I will not let this languish long, it's just non-essential to the goal of this current refactoring/pull request.

Reply:
:+1: